Remember dive lists in various charts

bstoeger opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
Remember the dives that are represented by bars, pie slices and box-and-whisker items. The goal is to make the corresponding charts selectable.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
1) Remember dives in the various chart data representations.

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->
Preparation for #3171.

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@atdotde: this is completely untested, but provides a basis for your PR #3171.

atdotde reviewed (approved):
Looks all good to me who still has to sit down and properly understand what std::move is good for.

Sorry for the delay.

bstoeger commented:
No worries - it's not a pressing issue to me.

> Looks all good to me who still has to sit down and properly understand what std::move is good for.

It's an unimportant implementation detail: It turns a reference into an r-value reference, which means that the caller is fine with the callee taking ownership of the data. When the function returns, the passed in vector may be empty ("moved from").

bstoeger commented:
Added a commit that replots the statistics if dives are added / removed.

This should solve the problem of having stale dive-pointers.

bstoeger commented:
@atdotde: I updated your commit to work with the other chart types and integrated it into this PR. I took the liberty to change from double-click to single-click. I'm not aware of any GUI where you select with double click? Moreover, I fixed the uninitialized pointer problem in the original commit.

There is still lots of work with respect to selection:
1) Holding the shift key should allow for additional selection.
2) There should be visual feedback concerning the selected items.
3) The interaction between selected items and selected dives is unclear to me. For example, when selecting dives from the list should the corresponding items in the chart be highlighted? What if only some of the dives of a bar are selected?

bstoeger commented:
This slightly improves the selection in the scatter-chart:
1) Shows selected dives. The color scheme is horrible, improvements are very welcome.
2) Allow multiple selections by shift-click.

I will try to implement similar modes for the other charts, but it will take some time.

bstoeger commented:
This now adds a rectangle "lasso" in scatter plots.

In combination with the shift-key, this is rather convenient, as different regions of the scatterplot can be selected.

Two sad points:
1) The list selection code is dog slow and the whole thing becomes useless when selecting a large number of dives. Probably the selection should only be performed when releasing the mouse button, but that adds a whole new level of complexity to the code
2) There is a funny bug when releasing the mouse button while on top of a scatter item: the info box will have a distorted size. That should be easy to fix (much easier than point 1, at least).

bstoeger commented:
Very slow progress here: highlight bars that are selected. The color-scheme is hideous, yes. Maybe someone has pity and comes up with nicer colors?

bstoeger commented:
More slow progress: support CTRL-based multi selection for all series.

dirkhh commented on stats/barseries.cpp:
```diff
@@ -152,6 +152,7 @@ void BarSeries::BarLabel::updatePosition(bool horizontal, bool center, const QRe
 				setVisible(false);
 				return;
 			}
+			pos.rx() -= itemSize.width() / 2.0;
```
didn't we merge this already?

dirkhh reviewed (commented):
This seriously makes my head hurt.
So much complexity - and how many people will use it. Ugh.
I can't find anything that's __wrong__ (except for the commit we already have), but I am not at all confident that it's all __right__, either.
I need to find time to test this on desktop (and to make sure there are no odd side effects on mobile where we don't support multiple selection today). Work is keeping me very busy, not sure I'll find time to do that today.

bstoeger commented:
Added shift-based range selection. Now there is a dichotomy between scatter plots and the other plots: the former supports "lasso" selection with a rectangle, the latter support range selection. Range selection is a bit weird for scatter plots. On the other hand, lasso selection might be useful for the others. But is it worth the hassle?

This also shows a problem of the selection code. The first time lots of dives are selected, this becomes extremely slow. We had a similar problem with CTRL-A. This will have to be investigated.

> So much complexity - and how many people will use it.

We had at least two requests on the mailing list and @atdotde who felt that it was important enough to do a first implementation. Therefore, I reckon that it is a surprisingly desired feature...?

The idea is of course to add a "restrict statistics to selection" feature, so that a subset of the data can be analyzed.

Also, this will have to be documented, especially the different selection modes with shift, control and lasso (paging @willemferguson).

dirkhh commented:
> Added shift-based range selection. Now there is a dichotomy between scatter plots and the other plots: the former supports "lasso" selection with a rectangle, the latter support range selection. Range selection is a bit weird for scatter plots. On the other hand, lasso selection might be useful for the others. But is it worth the hassle?

instinctive answer: no

> This also shows a problem of the selection code. The first time lots of dives are selected, this becomes extremely slow. We had a similar problem with CTRL-A. This will have to be investigated.

Yes, please. I thought we did away with the additional data structures that made this so slow. Are we still keeping additional state or additional lists of dives that we maintain?

> > So much complexity - and how many people will use it.
> 
> We had at least two requests on the mailing list and @atdotde who felt that it was important enough to do a first implementation. Therefore, I reckon that it is a surprisingly desired feature...?

I'm sure there'll be some people who'll use it. My thinking is more "we have 20k+ users, how many of them still predominantly use desktop? how many of those use statistics? how many of those will us the selection feature?"
 
> The idea is of course to add a "restrict statistics to selection" feature, so that a subset of the data can be analyzed.

Right. In order to stay sane... will you create a way to derive a filter criterion from the selection?
 
> Also, this will have to be documented, especially the different selection modes with shift, control and lasso (paging @willemferguson).

Yup. Also, this moves out the possible date for a release as we get this figured out and tested.
Not that's I'd have time for a release right now, anyway.

bstoeger commented:
> Yes, please. I thought we did away with the additional data structures that made this so slow. Are we still keeping additional state or additional lists of dives that we maintain?

It's the expansion of all trips in `DiveListView` that is ridiculously slow:
```
        for (int row: affectedTrips) {
                QModelIndex idx = m->index(row, 0);
                expand(idx);
        }
```

It is unclear whether this is a Qt issue or something we mishandle.

Once the trips are expanded, mass-selection is reasonably fast.

bstoeger commented:
Well ... a significant part of the problem seems to be that the location and dive icons are rendered for every single displayed dive:

```
                case LOCATION:
                        if (dive_has_gps_location(d)) {
                                IconMetrics im = defaultIconMetrics();
                                return QIcon(":globe-icon").pixmap(im.sz_small, im.sz_small);
                        }
                        break;
                case PHOTOS:
                        if (d->pictures.nr > 0) {
                                IconMetrics im = defaultIconMetrics();
                                return QIcon(icon_names[countPhotos(d)]).pixmap(im.sz_small, im.sz_small);
                        }        // If there are photos, show one of the three photo icons: fish= photos during dive;
                        break;   // sun=photos before/after dive; sun+fish=photos during dive as well as before/after
                }
                break;
```
This is not a particularly good idea.

bstoeger commented:
I implemented a very primitive "restrict to selection" interface. This works on top of the filter, because I wasn't sure how to implement it as a filter condition (how should it be saved)?

The UI is really ugly, but it works.

From my side, this is ready for user testing.

dirkhh commented:
So... the only way to get user testing is to merge into master. Is that what you are suggesting?

bstoeger commented:
This was mostly meant as an announcement that I'm done with this PR until I get change requests. I'm not sure if it is ready for merging as it is still rough around the corners .. the color-scheme...🤮

dirkhh commented:
The orange didn't bug me - the green did. So trivial commit added.
I still need to fix all this for dark mode... but I'm struggling to find mental space these days to dig into stuff like this.
Right now I'm playing with this PR, I'll see if I see other things that need fixing and otherwise I'll merge it.

dirkhh commented:
rebased, my commit hidden, completed another review. Ready to merge once the CI completes

bstoeger commented:
> I still need to fix all this for dark mode... but I'm struggling to find mental space these days to dig into stuff like this.

No worries. I will see if I can come up with some primitive theming machanism. It's not completely trivial, because some textures are cached, but it should also not be terribly difficult.

willemferguson commented:
On 2021/02/13 23:12, bstoeger wrote:
>
>     I still need to fix all this for dark mode... but I'm struggling
>     to find mental space these days to dig into stuff like this.
>
> No worries. I will see if I can come up with some primitive theming 
> machanism. It's not completely trivial, because some textures are 
> cached, but it should also not be terribly difficult.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/subsurface/subsurface/pull/3175#issuecomment-778678916>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABGX5MX27LN3ER2PHBUXU6TS63TKXANCNFSM4WKW2NJQ>.
>
I have a comment or 2.

1) The title of this option should be a bit more descriptive. 
Restriction of what? I would say "Restrict further analyses"

2) My immediate impulse is to reduce this tool from 4 vertical items to 
2 items:

     i) No heading equivalent to "Restriction"

     ii) a single-line checkbox with text "Restrict further analyses"

     ii) a line with the text equivalent to "Analysing subset (xx) 
dives". Possibly do not show any text when the whole dataset is being 
analysed? (i.e. do not show "Analysing all dives" if checkbox is not 
checked).

On my 1600 x 1200 screen the setup panel is already getting rather full 
of items and any further additions (after the present addition) will 
affect the vertical positioning of the different panels of the stats 
tool. Therefore the idea of minimising the vertical space requirement of 
the Restriction specification in the lefthand panel.

This PR makes the stats tool insanely flexible. I love it. Two questions:

1) Is this in a state where the documentation needs update?

2) I assume this will go to the mobile platform as well. Should I wait 
for generating the documentation for the mobile platform until this is 
finalised?

Kind regards,

wf




-- 
This message and attachments are subject to a disclaimer.

Please refer to 
http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf 
<http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf> for
full 
details.

bstoeger commented:
Concerning the desktop UI, I think the best way forward is that you make a pull request, as you did for the dive-detail tabs. That should avoid any miscommunications.

I'm not currently planning to make this fully work on mobile. It's not completely trivial to port, since mobile doesn't have "control" and "shift" keys. Moreover, I'm not familiar with mobile UIs, so I wouldn't even what the "right" thing to do is. Also note that the statistics on mobile are still incomplete: there is no way of turning on/off the features such as the median indicators.
